File: api/transcript.py
from youtube_transcript_api import YouTubeTranscriptApi
from youtube_transcript_api.formatters import TextFormatter
from urllib.parse import parse_qs, urlparse
import json

def extract_video_id(url):
    query = urlparse(url)
    if query.hostname in ["www.youtube.com", "youtube.com"]:
        return parse_qs(query.query).get("v", [None])[0]
    if query.hostname == "youtu.be":
        return query.path[1:]
    return None

def handler(request):
    video_url = request.args.get("url")
    if not video_url:
        return {
            "statusCode": 400,
            "body": json.dumps({"error": "Missing video URL"})
        }

    video_id = extract_video_id(video_url)
    if not video_id:
        return {
            "statusCode": 400,
            "body": json.dumps({"error": "Invalid YouTube URL"})
        }

    try:
        transcript = YouTubeTranscriptApi.get_transcript(video_id)
        formatter = TextFormatter()
        text = formatter.format_transcript(transcript)

        return {
            "statusCode": 200,
            "body": json.dumps({
                "video_id": video_id,
                "transcript": text
            })
        }

    except Exception as e:
        return {
            "statusCode": 500,
            "body": json.dumps({"error": str(e)})
        }

File: api/transcript.js
export default async function handler(req, res) {
  const videoId = req.query.videoId;
  if (!videoId) {
    return res.status(400).json({ error: "videoId missing" });
  }

  const r = await fetch(
    `https://youtubetranscript.p.rapidapi.com/?id=${videoId}`,
    {
      headers: {
        "X-RapidAPI-Key": process.env.RAPIDAPI_KEY,
        "X-RapidAPI-Host": "youtubetranscript.p.rapidapi.com"
      }
    }
  );

  const data = await r.json();
  res.status(200).json(data);
}
